### ISSUE TITLE
Rule Change: `possiblyUndefinedArray?.length === 0`

### ISSUE THREAD

**Issue by huntharo** [eslint/eslint#16562]
### What rule do you want to change?

no-unsafe-optional-chaining

### What change to do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
const noItemsReturned = possiblyUndefinedArray?.length === 0;

if (noItemsReturned) {
  console.log('we wanted to catch all cases where there is no array or no items, but we missed no array');
}
```

### What does the rule currently do for this code?

- No warning, no error
- The disallow arithmetic option doesn't list === as disallowed

### What will the rule do after it's changed?

- Warn when doing a strict `===` check after an optional chaining
- This *possibly* needs to be restricted to cases of `?.length === 0`, 
- Should encourage developers to change this to either:
  - `!possiblyUndefinedArray || possiblyUndefinedArray.length === 0`
  - `!possiblyUndefinedArray?.length`
  - other options?

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

I've seen a code base littered with this with comments saying "if no items show this `hey we didn't find anything` message"... but it conveniently misses the cases where the array is undefined (which is allowed / expected by types / other checks since the optional chaining operator is present).

I don't think any rule can catch this yet and it seems super dangerous and an easy mistake to make.

**Comment by mdjermanovic** [comment]
Hi @huntharo, thanks for the issue!

 I'm not sure if this check qualifies for `no-unsafe-optional-chaining`, as this rule warns about possible logic errors in code while `possiblyUndefinedArray?.length === 0` seems like a valid way to determine if `possiblyUndefinedArray` is an empty array. We can't know if the intent is to treat undefined the same as empty. Curious what other team members think.

> * This _possibly_ needs to be restricted to cases of `?.length === 0`

To catch this specific pattern, you can configure the [no-restricted-syntax](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule with the following selector:

```
BinaryExpression[operator='==='][right.value=0] > ChainExpression.left > MemberExpression.expression > Identifier.property[name='length']
```

[Playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXJlc3RyaWN0ZWQtc3ludGF4OiBbXCJlcnJvclwiLCB7XG4gICAgc2VsZWN0b3I6IFwiQmluYXJ5RXhwcmVzc2lvbltvcGVyYXRvcj0nPT09J11bcmlnaHQudmFsdWU9MF0gPiBDaGFpbkV4cHJlc3Npb24ubGVmdCA+IE1lbWJlckV4cHJlc3Npb24uZXhwcmVzc2lvbiA+IElkZW50aWZpZXIucHJvcGVydHlbbmFtZT0nbGVuZ3RoJ11cIixcbiAgICBtZXNzYWdlOiBcIlVuc2FmZSB1c2Ugb2Ygb3B0aW9uYWwgY2hhaW5pbmc6ICc/Lmxlbmd0aCA9PT0gMCcuXCJcbn1dICovXG5cbmNvbnN0IG5vSXRlbXNSZXR1cm5lZCA9IHBvc3NpYmx5VW5kZWZpbmVkQXJyYXk/Lmxlbmd0aCA9PT0gMDtcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnsiZXM2Ijp0cnVlfX19)

**Comment by nzakas** [comment]
> We can't know if the intent is to treat undefined the same as empty. Curious what other team members think.

I agree. I don’t think it’s clear that this pattern is problematic or that it should be flagged in this rule. A custom rule seems the best path forward if this pattern is important to you.
